prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+	/* Common info about ports */
+	int num_ports;
+	/* The list of ports that are active */
+	struct dlist_entry hw_ports_list;
+} ofi_verbs_ud_dev_ports_hw_t;
+
+typedef struct ofi_verbs_ud_device_hw_t {
+	struct dlist_entry list_entry;
+	struct ibv_context *ibv_ctx;
+	struct ibv_device_attr *attr;
+	struct ibv_qp_init_attr *qp_init_attr;
+	ofi_verbs_ud_dev_ports_hw_t *ports;
+} ofi_verbs_ud_device_hw_t;
+
+typedef struct ofi_verbs_ud_devices_hw_t {
+	/* Common info about devices */
+	struct ibv_context **ibv_ctx_list;
+	int num_devices;
+	/* The list of devices that are active */
+	struct dlist_entry devices_list;
+} ofi_verbs_ud_devices_hw_t;
+
+typedef struct ofi_verbs_ud_infra_hw_t {
+	ofi_verbs_ud_devices_hw_t *devices;
+} ofi_verbs_ud_infra_hw_t;
+
+/* ~All active HW devices structures */
+
+/* ~Init devices structures */
+
+int ofi_verbs_ud_fabric(struct fi_fabric_attr *attr,
+			struct fid_fabric **fabric_fid,
+			void *context);
+int ofi_verbs_ud_domain_open(struct fid_fabric *fabric_fid,
+			     struct fi_info *info,
+			     struct fid_domain **domain_fid,
+			     void *context);
+int ofi_verbs_ud_endpoint_open(struct fid_domain *domain_fid,
+			       struct fi_info *info,
+			       struct fid_ep **ep_fid,
+			       void *context);
+int ofi_verbs_ud_cq_open(struct fid_domain *domain_fid,
+			 struct fi_cq_attr *attr,
+			 struct fid_cq **cq_fid,
+			 void *context);
+int ofi_verbs_ud_av_open(struct fid_domain *domain_fid,
+			 struct fi_av_attr *attr,
+			 struct fid_av **av_fid, void *context);
+struct fi_ops_mr *ofi_verbs_ud_get_domain_mr_ops(void);
+struct fi_ops_msg *ofi_verbs_ud_get_ep_msg_ops(void);
+
+void ofi_verbs_ud_fini(void);
+
+/* Init functions */
+
+int ofi_verbs_ud_query_qp_attrs(struct ibv_context *ctx,
+				struct ibv_device_attr *device_attr,
+				struct ibv_qp_init_attr *init_attr);
+int ofi_verbs_ud_get_matching_info(uint32_t version, const char *dev_name,
+				   const struct fi_info *hints,
+				   struct fi_info **info,
+				   const struct fi_info *prov_info);
+
+static inline
+void ofi_verbs_ud_fork_init(void)
+{
+	int ret = FI_SUCCESS;
+
+	if (!gl_data.fork_unsafe) {
+		ret = ibv_fork_init();
+		if (ret) {
+			VERBS_UD_WARN(FI_LOG_CORE,
+				      "Enabling IB fork support failed: "
+				      "%s (%d)\n", strerror(ret), ret);
+		}
+		VERBS_UD_INFO(FI_LOG_CORE, "IB fork support "
+			      "is enabled\n");
+	} else {
+		VERBS_UD_INFO(FI_LOG_CORE, "IB fork support wasn't requested, "
+			      "not enabled it\n");
+	}
+}
+
+static inline
+int ofi_verbs_ud_have_device(void)
+{
+	struct ibv_device **devices;
+	struct ibv_context *verbs;
+	int found = 0, num_devices = 0, i;
+
+	devices = ibv_get_device_list(&num_devices);
+	if (!devices || !num_devices)
+		return found;
+
+	for (i = 0; i < num_devices; i++) {
+		verbs = ibv_open_device(devices[i]);
+		if (verbs) {
+			ibv_close_device(verbs);
+			found = 1;
+			break;
+		}
+	}
+
+	ibv_free_device_list(devices);
+	return found;
+}
```

Comment:
This function is just reporting if there is some verbs device present.  It doesn't guarantee that it can support UD.